### DISCUSSION TITLE
maxDuration and server actions

### DISCUSSION THREAD

**Discussion by eric-burel** [vercel/next.js#58855] [App Router]
Hi, I am trying to migrate some API routes towards Server Actions, but I am stuck with `maxDuration`: how can we set it for the POST endpoints generated by server actions?

**Comment by mordechaim** [comment]
You set it on the route segment that executes the action, i.e. the layout or page that does not have a `"use client"` directive.
https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#maxduration

**Reply by eric-burel** [reply]
Hi, it's probably right, but the documentation you linked doesn't confirm this, it only deals with routes, layouts and page.

**Reply by mordechaim** [reply]
A server action is considered a function executing on the page route, so they share the config.

**Reply by tillka** [reply]
I faced the same issue with my server actions timing out after 15s. Exporting maxDuration didn't work in the server action file itself so I moved it up to the next non-"use client" parent. However, I am getting the following error in typescript: 

`"maxDuration" is not a valid Next.js entry export value.ts(71002)`

**Reply by timothywadecook** [reply]
Hey @tillka I'm getting the same typescript error. Were you able to set maxDuration for server actions?

**Reply by ShayDavidson** [reply]
Getting the same problem. Upgraded to Pro plan especially to overcome the 10 seconds limitation, and I don't want to move my server actions to an API route.

**Reply by eric-burel** [reply]
Hi folks, you'd want to open a separate issue for this with repro, either on Vercel side or on Next side. I initially opened a discussion because it was a matter of documentation, but what you hit here is a bug so you need a ticket to get it fixed.

**Reply by tillka** [reply]
For sake of completness from my side: 
I included the maxDuration on page.tsx calling the sever action, as follows.
`export const maxDuration = 300; `
This throws the error `"maxDuration" is not a valid Next.js entry export value.ts(71002)` 
I put `// @ts-ignore` above the maxDuration statement - which doesn't remove the error in my VS Code but it successfully builds AND does the job as intended, i.e. overcoming the 15 seconds funcition limitation. Hope that helps

**Reply by ShayDavidson** [reply]
I eventually move the `maxDuration` to the server component page using the action, and then it worked for me. But this is not documented anywhere as far as I my search went, besides random mentions in Github Issues.

**Reply by evan-runloopai** [reply]
Is it expected that the vercel global config should still apply to server actions?
These [configuration docs](https://vercel.com/docs/functions/configuring-functions/duration#setting-a-default-maximum-duration) make it seem like you should be able to configure all server actions to run for a longer timeout of 60 seconds using a glob pattern like:
```
{
  "functions": {
    "app/**/*": {
      "maxDuration": 60
    }
  }
}
```

~~It seems to not apply when I test my app. I can open an issue report if it's actual bug but seems like maybe it's not documented that NextJS server actions have their separate configuration mechanism?~~


[Edit]:
Seemed I forgot I was using a src directory 😅 so my vercel.json line needed to be prefixed with 'src/' but I can confirm this configuration seemingly works if anyone is trying to change all their server action timeouts while they migrate from API routes

**Reply by bWgibb** [reply]
```
{
  "functions": {
    "src/app/**/*": {
      "maxDuration": 30
    }
  }
}
```
This worked for me to get my 15s server action to work (hobby plan).

**Comment by berkayermis** [comment]
I also was looking an explicit sentence in the docs but couldn't see anything.

So, I wanted to add the maxDuration at the most top level in my root page.tsx as @tillka did. Also, I upgraded my free plan to pro. Now, it works as I desired.

Version: Nextjs v14.1.0 with App Router.

**Reply by mordechaim** [reply]
Quoting the [docs](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#behavior)

> Server Actions inherit the Route Segment Config from the page or layout they are used on, including fields like `maxDuration`.

**Reply by tylerbecks** [reply]
> Quoting the [docs](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#behavior)
> 
> > Server Actions inherit the Route Segment Config from the page or layout they are used on, including fields like `maxDuration`.

This fixed it to me. I added the maxDuration on my page instead of the server function file.

**Comment by yaman-694** [comment]
I am using server action to call an api which takes upto 25s to response and db operations are perform in the backend.  I am on the pro plan which support 60s for the API routes. how to resolve this issue

**Reply by dryor** [reply]
Hey, @yaman-694 you can create vercel.json and configure the maximum duration here the durations for each plan https://vercel.com/docs/functions/runtimes#max-execution-time

**Comment by abhiamishra** [comment]
This issue seems to have opened up with the Hobby plan now:

https://github.com/orgs/vercel/discussions/7094#discussion-6781027

**Comment by Pratham271** [comment]
How can I increase the timeout for vercel server actions

**Comment by tylerbecks** [comment]
I tried everything, and only one thing worked for me. Note: I'm on React 19 and NextJS 15.

Adding this to my page component from where the server action is called fixed it:
```ts
export const maxDuration = 300;
```

What didn't work:
- Trying to configure `vercel.json`. I tried many iterations of this and nothing worked, [the docs](https://vercel.com/docs/projects/project-configuration#description) seem outdated because they still mention the pages/ router instead of the app router. The docs also mention that the file must be in the `api/` directory, which doesn't seem to match the NextJS docs on server components
- Setting maxDuration in the server function failed because I have the `'use server'` directive on the top of the and it says I can only export async functions from files with the `'use server'` directive.
- Adding @ts-ignore didn't work for me

Overall, I'm disappointed in the state of the documentation from Vercel on server actions.

**Reply by SokratisVidros** [reply]
I’m experiencing the same issue, setting `maxDuration` with `use server` still isn’t working and is very unintuitive.

Additionally, it would be really helpful if Next.js allowed configuring a global maxDuration in `next.config.js`, especially for projects not deployed on Vercel. Overriding the setting on a per-file basis completes the puzzle.
